From 0048e125ae7f45dc9283896dda9ecc1690b3049a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Vincent=20Harl=C3=A9?= <vincent.harle@softathome.com>
Date: Tue, 13 Sep 2022 14:35:50 +0200
Subject: [PATCH] scripts: environment: fix DeviceInfo.SerialNumber
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Currently TR69 and LCM are failing due to empty DeviceInfo.SerialNumber
variable, probably due to the race condition with br-lan interface and
MAC address retrievel. So lets fix it by trying to retrieve the MAC
address from br-lan interface for 3 seconds and use dummy MAC address if
it takes longer.

References: PCF-660
Fixes: fc4c2e4e1c24 ("Issue: HOP-1753 [DeviceInfo][NEC WX3000HP] TR69 and LCM failing due to empty DeviceInfo.SerialNumber [fix]")
Signed-off-by: Vincent Harlé <vincent.harle@softathome.com>
Signed-off-by: Petr Štetiar <ynezz@true.cz> [adjusted copy&pasted script]
Signed-off-by: Petr Štetiar <ynezz@true.cz>
---
 scripts/environment.sh | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/scripts/environment.sh b/scripts/environment.sh
index 5bb9e00111ac..9b6273f5899a 100644
--- a/scripts/environment.sh
+++ b/scripts/environment.sh
@@ -1,17 +1,32 @@
-#DEFAULT VALUES FOR POPULATING DEVICEINFO
+#GENERIC SCRIPT FOR POPULATING DEVICEINFO
+#This script should be customized for specific device whit proper source of manufacturing data
 export DEVICECATEGORY=""
 export CID="000000"
 export PEN=""
 
+#Generic approach to find MAC address from br-lan interface to build a default OUI and SerialNumber
+#Some Device dynamicaly sets the MAC address and this takes time to be applied.
+#Hence we wait for 3 seconds and default to a dummy MAC if takes longer.
+MACADDRESS=01:00:00:12:34:56;
+for i in 1 2 3 ; do
+	ADDR=$(cat /sys/class/net/br-lan/address)
+	if [ -n "$ADDR" ] ; then
+		MACADDRESS=$ADDR
+		break;
+	fi
+	sleep 1
+done
+
+
 #VALUES USING BASH
 export UPTIME=$(awk '{ printf("%i", $1) }' /proc/uptime)
 export MANUFACTURER=$(awk -F',' '{print $1}' /tmp/sysinfo/board_name)
 export PRODUCTCLASS=$(awk -F',' '{print $2}' /tmp/sysinfo/board_name)
-export MANUFACTUREROUI=$(cat /sys/class/net/br-lan/address | tr -d ':' | head -c 6 | tr '[a-z]' '[A-Z]')
+export MANUFACTUREROUI=$(echo $MACADDRESS | tr -d ':' | head -c 6 | tr '[a-z]' '[A-Z]')
 export MODELNAME=$(cat /tmp/sysinfo/model)
 export MODELNUMBER=""
 export DESCRIPTION=$(awk '$1 ~ /DISTRIB_DESCRIPTION/' /etc/openwrt_release | cut -d "'" -f2 | cut -d "'" -f1)
-export SERIALNUMBER=SN$(cat /sys/class/net/br-lan/address | tr -d ':')
+export SERIALNUMBER=SN$(echo $MACADDRESS | tr -d ':')
 export HARDWAREVERSION=$(awk '$1 ~ /DISTRIB_TARGET/' /etc/openwrt_release | cut -d "'" -f2 | cut -d "'" -f1)
 export SOFTWAREVERSION=$(grep -oE 'Linux version [0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}' /proc/version | awk '{ print $NF }')
 export TOTALMEMORY=$(awk '$1 ~ /MemTotal/{ print $2 }' /proc/meminfo)
